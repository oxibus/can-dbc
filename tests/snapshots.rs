#![cfg(all(feature = "serde", feature = "encodings"))]

use nom::Err;
use std::borrow::Cow;
use std::fs;
use std::path::{Path, PathBuf};

use can_dbc::{decode_cp1252, Dbc, Error};
use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use test_each_file::test_each_path;

struct TestConfig {
    test_root: &'static str,
    snapshot_suffix: &'static str,
    use_cp1251: bool,
    create_snapshot: bool,
}

static TEST_DIRS: &[TestConfig] = &[
    TestConfig {
        test_root: "shared-test-files",
        snapshot_suffix: "",
        use_cp1251: true,
        create_snapshot: true,
    },
    TestConfig {
        test_root: "opendbc/opendbc/dbc",
        snapshot_suffix: "opendbc",
        use_cp1251: false,
        create_snapshot: false,
    },
];

test_each_path! { for ["dbc"] in "./tests/fixtures/opendbc/opendbc/dbc" as dbc => parse_one_file }
test_each_path! { for ["dbc"] in "./tests/fixtures/shared-test-files" as shared => parse_one_file }
// upper case extension
test_each_path! { for ["DBC"] in "./tests/fixtures/shared-test-files" as shared2 => parse_one_file }

/// Get `test root`, `snapshot name suffix`, `use cp1251`, `create snapshot` for the given path
fn get_test_info(path: &Path) -> Option<(PathBuf, &'static TestConfig)> {
    if !path
        .extension()
        .unwrap_or_default()
        .eq_ignore_ascii_case("dbc")
    {
        return None;
    }
    let path_str = path.to_str().unwrap();
    let parent = path.parent().unwrap();
    for item in TEST_DIRS {
        // Ensure slashes are there for easier matching
        let test_root = format!("/{}/", item.test_root);
        let mut path_dir = parent.to_str().unwrap().to_string();
        if !path_dir.ends_with('/') {
            path_dir.push('/');
        }
        if let Some(pos) = path_dir.find(&test_root) {
            let parent = PathBuf::from("snapshots")
                .join(item.snapshot_suffix)
                .join(&path_dir[pos + test_root.len()..]);
            return Some((parent, item));
        }
    }
    panic!("Unknown test directory: {path_str}");
}

/// Test parsing all DBC files
#[test]
fn test_if_submodules_are_present() {
    for test in TEST_DIRS {
        let dir = Path::new("./tests/fixtures").join(test.test_root);
        fs::read_dir(&dir).unwrap_or_else(|e| {
            let dir_display = dir.display();
            panic!(
                "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
            )
        });
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file([path]: [&Path; 1]) {
    let Some((
        snapshot_path,
        &TestConfig {
            use_cp1251,
            create_snapshot,
            ..
        },
    )) = get_test_info(path)
    else {
        return;
    };

    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if use_cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    report_test_result(
        Dbc::try_from(buffer.as_ref()),
        snapshot_path,
        file_name,
        create_snapshot,
    );
}

fn report_test_result(
    result: Result<Dbc, Error>,
    snapshot_path: PathBuf,
    file_name: String,
    do_snapshot: bool,
) {
    let do_snapshot = std::env::var("SKIP_INSTA").is_err()
        && (do_snapshot || std::env::var("FORCE_INSTA").is_ok());

    with_settings! {
        {
            omit_expression => true,
            snapshot_path => snapshot_path,
            prepend_module_to_snapshot => false
        },
        {
            match result {
                Ok(result) => {
                    if do_snapshot {
                        assert_yaml_snapshot!(file_name, result);
                    } // else ignore successful results when not snapshotting
                }
                Err(e) => {
                    let file_name = format!("!error___{file_name}");
                    assert_debug_snapshot!(file_name, err_to_str(&e));
                }
            }
        }
    }
}

fn err_to_str<'a>(err: &'a Error) -> &'a str {
    match err {
        Error::Incomplete(_, _) => "incomplete",
        Error::Nom(e) => match e {
            Err::Incomplete(_) => "nom-incomplete",
            Err::Error(_) => "nom-error",
            Err::Failure(_) => "nom-failure",
        },
        Error::MultipleMultiplexors => "multiple multiplexors",
    }
}
