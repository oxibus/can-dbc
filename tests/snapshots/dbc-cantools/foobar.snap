---
source: tests/snapshots.rs
---
version: "2.0"
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing: []
nodes:
  - FOO
  - BAR
  - FIE
  - FUM
value_tables:
  - name: State
    descriptions:
      - id: 1
        description: Enabled
      - id: 0
        description: Disabled
messages:
  - id:
      Extended: 74544
    name: Foo
    size: 8
    transmitter:
      NodeName: FOO
    signals:
      - name: Foo
        multiplexer_indicator: Plain
        start_bit: 0
        size: 12
        byte_order: BigEndian
        value_type: Signed
        factor: 0.01
        offset: 250
        min: 229.53
        max: 270.47
        unit: degK
        receivers:
          - BAR
      - name: Bar
        multiplexer_indicator: Plain
        start_bit: 24
        size: 32
        byte_order: BigEndian
        value_type: Signed
        factor: 0.1
        offset: 0
        min: 0
        max: 5
        unit: m
        receivers:
          - FOO
  - id:
      Extended: 74545
    name: Fum
    size: 5
    transmitter:
      NodeName: FOO
    signals:
      - name: Fum
        multiplexer_indicator: Plain
        start_bit: 0
        size: 12
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 10
        unit: ""
        receivers:
          - BAR
      - name: Fam
        multiplexer_indicator: Plain
        start_bit: 12
        size: 12
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 8
        unit: ""
        receivers:
          - BAR
  - id:
      Extended: 74546
    name: Bar
    size: 4
    transmitter:
      NodeName: FOO
    signals:
      - name: Binary32
        multiplexer_indicator: Plain
        start_bit: 0
        size: 32
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - FUM
  - id:
      Extended: 74547
    name: CanFd
    size: 64
    transmitter:
      NodeName: FOO
    signals:
      - name: Fie
        multiplexer_indicator: Plain
        start_bit: 0
        size: 64
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - FUM
      - name: Fas
        multiplexer_indicator: Plain
        start_bit: 64
        size: 64
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - id:
      Standard: 780
    name: FOOBAR
    size: 8
    transmitter:
      NodeName: FIE
    signals:
      - name: ACC_02_CRC
        multiplexer_indicator: Plain
        start_bit: 0
        size: 12
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - BAR
message_transmitters:
  - message_id:
      Extended: 74546
    transmitter:
      - NodeName: FOO
      - NodeName: BAR
comments:
  - Node:
      name: BAR
      comment: "fam \\\"1\\\""
  - Message:
      id:
        Extended: 74544
      comment: Foo.
  - Signal:
      message_id:
        Extended: 74544
      name: Bar
      comment: Bar.
attribute_definitions:
  - Plain: "\"DBName\" STRING"
  - Plain: "\"Baudrate\" INT 0 1000000"
  - Plain: "\"AFloat\" FLOAT 0 100"
  - Plain: "\"BusType\" STRING"
  - Node: "\"FOO\" INT 0 100"
  - Node: "\"BAR\" STRING"
  - Message: "\"VFrameFormat\" ENUM  \"StandardCAN\",\"ExtendedCAN\",\"reserved\",\"reserved\",\"reserved\",\"reserved\",\"reserved\",\"reserved\",\"reserved\",\"reserved\",\"reserved\",\"reserved\",\"reserved\",\"reserved\",\"StandardCAN_FD\",\"ExtendedCAN_FD\""
  - Message: "\"GenMsgStartValue\" STRING"
  - Message: "\"GenMsgCycleTime\" INT 0 1000"
relation_attribute_definitions:
  - name: AFloat2
    value_spec: "\"AFloat2\" FLOAT 0 65535"
  - name: GenSigTimeoutTime
    value_spec: "\"GenSigTimeoutTime\" INT 0 65535"
  - name: GenSigRespComment
    value_spec: "\"GenSigRespComment\" STRING"
attribute_defaults:
  - name: DBName
    value:
      String: ""
  - name: Baudrate
    value:
      Double: 125000
  - name: AFloat
    value:
      Double: 0
  - name: BusType
    value:
      String: ""
  - name: FOO
    value:
      Double: 0
  - name: BAR
    value:
      String: ""
  - name: VFrameFormat
    value:
      String: ExtendedCAN_FD
  - name: GenMsgStartValue
    value:
      String: ""
  - name: GenMsgCycleTime
    value:
      Double: 100
relation_attribute_defaults:
  - name: AFloat2
    value:
      Double: 0
  - name: GenSigTimeoutTime
    value:
      Double: 0
attribute_values:
  - name: DBName
    value:
      Raw:
        String: TheBusName
  - name: Baudrate
    value:
      Raw:
        Double: 125000
  - name: AFloat
    value:
      Raw:
        Double: 33.5
  - name: BusType
    value:
      Raw:
        String: CAN FD
  - name: BAR
    value:
      NetworkNode:
        - FIE
        - String: FUM
  - name: FOO
    value:
      NetworkNode:
        - FIE
        - Double: 1
  - name: VFrameFormat
    value:
      MessageDefinition:
        - Extended: 74544
        - Double: 1
  - name: VFrameFormat
    value:
      MessageDefinition:
        - Extended: 74545
        - Double: 1
  - name: GenMsgCycleTime
    value:
      MessageDefinition:
        - Extended: 74545
        - Double: 1
  - name: VFrameFormat
    value:
      MessageDefinition:
        - Extended: 74546
        - Double: 1
  - name: VFrameFormat
    value:
      MessageDefinition:
        - Extended: 74547
        - Double: 15
  - name: VFrameFormat
    value:
      MessageDefinition:
        - Standard: 780
        - Double: 0
relation_attribute_values:
  - name: GenSigTimeoutTime
    details:
      NodeToSignal:
        node_name: FOO
        message_id:
          Extended: 74544
        signal_name: Bar
        value:
          Double: 200
  - name: AFloat2
    details:
      NodeToSignal:
        node_name: FOO
        message_id:
          Extended: 74544
        signal_name: Bar
        value:
          Double: 500
value_descriptions:
  - Signal:
      message_id:
        Extended: 74545
      name: Fam
      value_descriptions:
        - id: 1
          description: Enabled
        - id: 0
          description: Disabled
signal_groups:
  - message_id:
      Extended: 74545
    name: SigGroupName
    repetitions: 1
    signal_names:
      - Fam
      - Fum
signal_extended_value_type_list:
  - message_id:
      Extended: 74544
    signal_name: Bar
    signal_extended_value_type: IEEEfloat32Bit
  - message_id:
      Extended: 74546
    signal_name: Binary32
    signal_extended_value_type: IEEEfloat32Bit
